### ISSUE TITLE
jobs: using pty=1 with buffered output will not close stdin/exit job

### ISSUE THREAD

**Issue by blueyed** [neovim/neovim#9107]
```vim
function! F(...) dict
  echom string(a:000).', '.string(self)
  echom string(s:opts)
endfunction

let opts = {}
let opts.stdout_buffered = 1
let opts.stderr_buffered = 1
let opts.on_exit = function('F')
" let opts.on_stdout = function('F')
" let opts.on_stderr = function('F')
let opts.pty = 1

let s:opts = opts

let job = jobstart(['cat', '/dev/stdin'], opts)
" let job = jobstart(['echo', '/dev/stdin'], s:opts)
call jobsend(job, ['foo'])
call jobclose(job, 'stdin')
```

Running this with `nvim -u t-dev-stdin-buffered.vim` will cause the job to not being closed/stopped.

The log says:

> INFO  2018-10-11T15:56:16.756 28066 channel_create_event:208: new channel 3 (…/t-dev-stdin-buffered.vim:16) : {"id": 3, "mode": "bytes", "stream": "job", "pty": "/dev/pts/25"}
> INFO  2018-10-11T15:56:16.824 28066 main:590: starting main loop

Calling `jobstop(job)` then says:
> INFO  2018-10-11T15:56:57.007 28066 on_process_exit:385: exited: pid=28067 status=1 stoptime=0

**Comment by blueyed** [comment]
This happens without buffered output also already:
```vim
function! F(...) dict
  echom string(a:000).', '.string(self)
  echom string(s:opts)
endfunction

let opts = {}
let opts.pty = 1

let s:opts = opts

let job = jobstart(['cat', '/dev/stdin'], opts)
call jobclose(job, 'stdin')
```

**Comment by blueyed** [comment]
Ah, it seems to be caused by not connecting `stdout` (not using `let opts.on_stdout = function('F')`), and not handling/reading `stdout` with buffered output correctly.

**Comment by blueyed** [comment]
Manually sending SIGHUP makes it work, but the exit code will be 1 then:

```vim
function! F(...) dict
  echom string(a:000).', '.string(self)
  echom string(s:opts)
endfunction

let opts = {}
let opts.stdout_buffered = 1
let opts.stderr_buffered = 1
let opts.on_exit = function('F')
" let opts.on_stdout = function('F')
" let opts.on_stderr = function('F')
let opts.pty = 1

let s:opts = opts

let job = jobstart(['cat', '/dev/stdin'], opts)
" let job = jobstart(['echo', '/dev/stdin'], s:opts)
call jobsend(job, ['foo'])
call jobclose(job, 'stdin')

call system('kill -SIGHUP '.jobpid(job))
```

Log:
```
INFO  2018-10-11T16:11:46.964 32436 channel_create_event:208: new channel 3 (…/t-dev-stdin-buffered.vim:16) : {"id": 3, "mode": "bytes", "stream": "job", "pty": "/dev/pts/25"}
INFO  2018-10-11T16:11:46.967 32436 on_process_exit:385: exited: pid=32437 status=1 stoptime=0
INFO  2018-10-11T16:11:46.967 32436 on_process_exit:385: exited: pid=32438 status=0 stoptime=0
INFO  2018-10-11T16:11:47.009 32436 main:590: starting main loop
```

`:mess`:
```
[3, 1, 'exit'], {'stdout_buffered': 1, 'on_exit': function('F'), 'stderr_buffered': 1, 'stdou
t': ['foo'], 'pty': 1}
```

**Comment by blueyed** [comment]
And with `pty=1` it appears to get `stdin` on `stdout`?!

The following will have "foo" on "stdout", but not "sbb"?!

```vim
function! F(...) dict
  echom string(a:000).', '.string(self)
endfunction

let opts = {}
let opts.on_exit = function('F')
let opts.on_stdout = function('F')
let opts.on_stderr = function('F')
let opts.pty = 1

let job = jobstart(['rot13'], opts)
call jobsend(job, ['foo'])
call jobclose(job, 'stdin')
```

`:mess`:
```
[3, ['foo'], 'stdout'], {'on_exit': function('F'), 'on_stdout': function('F'), 'on_stderr': f
unction('F'), 'pty': 1}
```

**Comment by bfredl** [comment]
Pty means stdin and stdout is the same fd. Use ctrl-d `\x04` to signal end of input to the child process.

**Comment by blueyed** [comment]
Ah.

> Use ctrl-d \x04 to signal end of input to the child process.

Wouldn't that be what `jobclose(job, 'stdin')` should do then?

As for the previous comment.. using `jobsend(job, nr2char(4))` then makes it stop, but it will have both "foo" and "sbb" on stdout then.
But that reflects the "stdin and stdout is the same fd" then.

It looks like "pty" and "stdin" cannot be used then together, right?

**Comment by blueyed** [comment]
What I wanted to achieve is having `/dev/stdin` (for hadolint) available, which appears to only be available with `pty=1`.

**Comment by bfredl** [comment]
> Wouldn't that be what jobclose(job, 'stdin') should do then?

No, because it doesn't close the (bidirectional) pty. It just finishes the current/next `read()` command the child process  does. Just as when you type CTRL-D at `cat`, it either finishes the current input without NL or exists `cat` if there is no pending input, but it doesn't (necessarily) close the terminal window but allows it to return to the shell.

> As for the previous comment.. using jobsend(job, nr2char(4)) then makes it stop, but it will have both "foo" and "sbb" on stdout then.

Yes. Compare what happens if you run `rot13` in a terminal window.

The pty could be configured to (1) not wait for NL or CTRL-D and (2) not echo the input back. A `stty` call should do this, I'll get back to it.

> What I wanted to achieve is having /dev/stdin (for hadolint) available, which appears to only be available with pty=1.

This is an unfortunate limitation. Only a subset of linux IPT is reified as "real" filenames that `/dev/stdin` can link to.

**Comment by blueyed** [comment]
This will turn off echo by default, which appears to be sensible, isn't it?

```diff
diff --git i/src/nvim/os/pty_process_unix.c w/src/nvim/os/pty_process_unix.c
index bafbfe1e4..a55f1e59d 100644
--- i/src/nvim/os/pty_process_unix.c
+++ w/src/nvim/os/pty_process_unix.c
@@ -57,6 +57,7 @@ int pty_process_spawn(PtyProcess *ptyproc)
   if (!termios_default.c_cflag) {
     // TODO(jkeyes): We could pass NULL to forkpty() instead ...
     init_termios(&termios_default);
+    termios_default.c_lflag &= ~(ECHO);
   }
 
   int status = 0;  // zero or negative error code (libuv convention)
```

How is "not wait for NL or CTRL-D" configured?
Also some generic function like `cfmakeraw` might be sensible?

**Comment by bfredl** [comment]
> This will turn off echo by default, which appears to be sensible, isn't it?

Why not leave the system defaults to their expected values and letting the user configure it if needed? Creating a set of nvim special defaults just adds yet a layer of arbitrary decisions an API user is expected to know about _in addition to_ knowing how PTY:s work in general, and knowing what the child does. Note as PTY childs can change these settings (as stty does), so a nvim default does not guarantee that these behaviours are disabled. It just means knowledge of default behavior is now not valid anymore.

**Comment by blueyed** [comment]
Yeah, that was only meant as some "this works in this case" patch - likely it would also affect `:term` in this case?!

Anyway, isn't `init_termios` biased anyway already?
https://github.com/neovim/neovim/blob/384770556ba02433904baf78f499b9bb48010040/src/nvim/os/pty_process_unix.c#L186-L239

**Comment by bfredl** [comment]
It might be, but in general the intention seems to be to be close to an ordinary terminal (standard echo behavior, cooked mode, standard key values), perhaps paper over some differences between different systems, but not to create nvim-specific behaviors as far as I can see.

**Comment by blueyed** [comment]
>> What I wanted to achieve is having /dev/stdin (for hadolint) available, which appears to only be available with pty=1.

> This is an unfortunate limitation. Only a subset of linux IPC is reified as "real" filenames that /dev/stdin can link to.

Where is this done?
How can this be simulated, e.g. to test `hadolint` outside of Neovim with this? (https://github.com/hadolint/hadolint/issues/273)

**Comment by bfredl** [comment]
> Where is this done?

`/dev/stdin` links to `/proc/self/fd/0`, and the later is a link set by the kernel, but not always to a real file name, for example:

```
% echo x | ls -l /proc/self/fd/
total 0
lr-x------ 1 bjorn bjorn 64 Oct 12 09:01 0 -> 'pipe:[49863]'
lrwx------ 1 bjorn bjorn 64 Oct 12 09:01 1 -> /dev/pts/1
lrwx------ 1 bjorn bjorn 64 Oct 12 09:01 2 -> /dev/pts/1
lr-x------ 1 bjorn bjorn 64 Oct 12 09:01 3 -> /proc/2019/fd
```
so anything like `cat filename| hadolint ...` should work to emulate the same situation (but not `hadolint /dev/stdin < filename`, because this will open the file directly and thus give the real file name).

Adding a `-` argument as in the linked issue would be the proper solution.

**Comment by blueyed** [comment]
Thanks again!

> `cat filename| hadolint ...`

That unfortunately does not work (directly) when using a wrapper script to run hadolint in a Docker image (`exec /usr/bin/docker run --rm -i hadolint/hadolint hadolint "${@:--}"`).

**Comment by bfredl** [comment]
Why not? does the container not contain bash or similar shell?

**Comment by bfredl** [comment]
The point here was to answer the question "How can this be simulated, e.g. to test hadolint outside of Neovim with this?", not solving the actual issue. But if that's still a question, `cat filename| /usr/bin/docker ...` I guess should work for this.

**Comment by blueyed** [comment]
Ah.. with `cat Dockerfile | hadolint -` `/dev/stdin` is available though.  I've thought that it would be related to how Neovim spawns the job, and could be simluated by some terminal option etc.
I think it is fine for now, and I'll wait to feedback on the hadolint issue to use the fd directly.

**Comment by blueyed** [comment]
Gave it a shot in https://github.com/neovim/neovim/pull/10557.

**Comment by shazow** [comment]
I was trying to put together a way to pipe my buffer into a binary and render the result in a terminal, but turns out not being able to control the echo ruins the use case, so I'm pretty excited about this issue! :)

My specific usecase (not dissimilar to the examples mentioned above):

```vim
let bufid = bufnr('%') | new | let id = termopen('mdcat') | call chansend(id, getbufline(bufid, 1, '$') + ["\r", "\004"])
```

mdcat takes some markdown and renders it using a bunch of ansi escape codes in a terminal. Right now I get the stdin echo prefacing mdcat's output which is unfortunate and difficult to work around. Tried to do all kinds of `stty -echo` trickery, but wasn't able to make it work nicely. This issue would make this scenario much nicer. :)
